File: sdk/android/frak-sdk/src/main/kotlin/id/frak/sdk/core/FrakConfig.kt
package id.frak.sdk.core

/** Currency a reward is advertised in. Closed set: backend rejects anything else with a 422. */
public enum class FrakCurrency(
    public val wireValue: String,
) {
    EUR("eur"),
    USD("usd"),
    GBP("gbp"),
}

/** Language for merchant-configured copy. Only `en`/`fr` exist today. */
public enum class FrakLanguage(
    public val wireValue: String,
) {
    EN("en"),
    FR("fr"),
}

/** Logcat verbosity. Default [NONE]. Also gates [FrakConfig.logSink] volume, see [FrakLogSink]. */
public enum class FrakLogLevel {
    NONE,
    ERROR,
    WARN,
    INFO,
    DEBUG,
}

/** Receives SDK diagnostics, gated by [FrakConfig.logLevel]. Replaces logcat once set. */
public fun interface FrakLogSink {
    /** Must not throw (exception is swallowed, not surfaced) and must be thread-safe. */
    public fun log(
        level: FrakLogLevel,
        message: String,
        throwable: Throwable?,
    )
}

/** How inbound links carrying an `fCtx` reach the SDK. */
public enum class DeepLinkHandling {
    /**
     * The SDK watches host activities itself; calling [id.frak.sdk.AppLinkApi.handleReferral]
     * as well double-tracks the arrival. Android-only.
     */
    Automatic,

    /** Merchant calls [id.frak.sdk.AppLinkApi.handleReferral] from their own router. */
    Manual,

    Disabled,
}

/**
 * Static merchant-supplied facts, fixed at build time. Not the resolved backend config, see
 * [id.frak.sdk.config.FrakResolvedConfig].
 */
public class FrakMetadata internal constructor(
    public val name: String?,
    public val currency: FrakCurrency,
    /** Null means "let the backend decide" (falls back to `en`). */
    public val lang: FrakLanguage?,
    public val logoUrl: String?,
    public val homepageLink: String?,
) {
    public class Builder {
        public var name: String? = null

        public var currency: FrakCurrency = FrakCurrency.EUR

        public var lang: FrakLanguage? = null

        public var logoUrl: String? = null

        public var homepageLink: String? = null

        public fun name(name: String?): Builder = apply { this.name = name }

        public fun currency(currency: FrakCurrency): Builder = apply { this.currency = currency }

        public fun lang(lang: FrakLanguage?): Builder = apply { this.lang = lang }

        public fun logoUrl(logoUrl: String?): Builder = apply { this.logoUrl = logoUrl }

        public fun homepageLink(homepageLink: String?): Builder = apply { this.homepageLink = homepageLink }

        public fun build(): FrakMetadata = FrakMetadata(name, currency, lang, logoUrl, homepageLink)
    }
}

/** Kotlin sugar over [FrakMetadata.Builder]. */
public fun FrakMetadata(configure: FrakMetadata.Builder.() -> Unit): FrakMetadata =
    FrakMetadata.Builder().apply(configure).build()

/**
 * Everything the SDK needs to start, supplied once at [id.frak.sdk.Frak.initialize]. Never
 * validated at construction; an unusable config surfaces later as [FrakError.MerchantResolutionFailed].
 */
public class FrakConfig internal constructor(
    /** Optional; when null, merchant is resolved from [packageId] instead. `merchantId` wins if both set. */
    public val merchantId: String?,
    /** Null reads `context.packageName` at [id.frak.sdk.Frak.initialize]. `bundleId` on iOS. */
    public val packageId: String?,
    public val metadata: FrakMetadata,
    /** Merchants never set this; exists for Frak's own dev/local builds. */
    public val env: FrakEnvironment,
    public val deepLink: DeepLinkHandling,
    /**
     * Hard floor for tracking that [id.frak.sdk.FrakClient.setTrackingEnabled] cannot lift at
     * runtime. `false` also stops sharing, but config and reward resolution still run.
     */
    public val trackingEnabled: Boolean,
    public val logLevel: FrakLogLevel,
    public val logSink: FrakLogSink?,
) {
    // See the note atop sharing/SharingRequest.kt.

    /**
     * `Builder()` exists alongside `Builder(merchantId)` because [merchantId] is optional. The empty
     * one is primary: a shared `constructor(String?)` would erase to the same JVM descriptor as
     * `constructor(String)`.
     */
    public class Builder() {
        public constructor(merchantId: String) : this() {
            this.merchantId = merchantId
        }

        public var merchantId: String? = null

        public var packageId: String? = null

        public var metadata: FrakMetadata = FrakMetadata.Builder().build()

        public var env: FrakEnvironment = FrakEnvironment.Production

        public var deepLink: DeepLinkHandling = DeepLinkHandling.Automatic

        public var trackingEnabled: Boolean = true

        public var logLevel: FrakLogLevel = FrakLogLevel.NONE

        public var logSink: FrakLogSink? = null

        public fun merchantId(merchantId: String?): Builder = apply { this.merchantId = merchantId }

        public fun packageId(packageId: String?): Builder = apply { this.packageId = packageId }

        public fun metadata(metadata: FrakMetadata): Builder = apply { this.metadata = metadata }

        public fun env(env: FrakEnvironment): Builder = apply { this.env = env }

        public fun deepLink(deepLink: DeepLinkHandling): Builder = apply { this.deepLink = deepLink }

        public fun trackingEnabled(trackingEnabled: Boolean): Builder = apply { this.trackingEnabled = trackingEnabled }

        public fun logLevel(logLevel: FrakLogLevel): Builder = apply { this.logLevel = logLevel }

        public fun logSink(logSink: FrakLogSink?): Builder = apply { this.logSink = logSink }

        public fun build(): FrakConfig =
            FrakConfig(
                merchantId,
                packageId,
                metadata,
                env,
                deepLink,
                trackingEnabled,
                logLevel,
                logSink,
            )
    }

    /** Not a `data class`: publishing one bakes `copy()`/`componentN()` into the ABI permanently. */
    internal fun withPackageId(packageId: String): FrakConfig =
        FrakConfig(
            merchantId = merchantId,
            packageId = packageId,
            metadata = metadata,
            env = env,
            deepLink = deepLink,
            trackingEnabled = trackingEnabled,
            logLevel = logLevel,
            logSink = logSink,
        )
}

/** Kotlin sugar over [FrakConfig.Builder], for the merchant-id form. */
public fun FrakConfig(
    merchantId: String,
    configure: FrakConfig.Builder.() -> Unit,
): FrakConfig = FrakConfig.Builder(merchantId).apply(configure).build()

/** Merchant id only: the shortest working config. */
public fun FrakConfig(merchantId: String): FrakConfig = FrakConfig.Builder(merchantId).build()

/** Kotlin sugar over [FrakConfig.Builder] for the no-merchant-id form. */
public fun FrakConfig(configure: FrakConfig.Builder.() -> Unit): FrakConfig =
    FrakConfig.Builder().apply(configure).build()

File: sdk/ios/Sources/FrakSDK/Core/FrakConfig.swift
// The two public `Decodable`s the read models deliberately don't get: here the conformance is just
// `rawValue`, which is already public and already the form we send on requests, so it freezes nothing.
public enum FrakCurrency: String, Sendable, CaseIterable, Decodable, Hashable {
    case eur
    case usd
    case gbp
}

public enum FrakLanguage: String, Sendable, CaseIterable, Decodable, Hashable {
    case en
    case fr
}

// Default .none is silent. Also gates FrakConfig.logSink, same as os.Logger.
public enum FrakLogLevel: Int, Sendable, Hashable, Comparable {
    case none
    case error
    case warn
    case info
    case debug

    public static func < (lhs: FrakLogLevel, rhs: FrakLogLevel) -> Bool {
        lhs.rawValue < rhs.rawValue
    }
}

public enum DeepLinkHandling: Sendable, Hashable {
    // Merchant calls FrakClient.appLink.handleReferral(_:) from onOpenURL or their own router.
    //
    // The only mode iOS offers: nothing lets a library install itself in front of the app's own
    // URL routing, so this SDK cannot intercept `onOpenURL`/`onContinueUserActivity` on the
    // merchant's behalf the way Android's `.automatic` mode can via `ActivityLifecycleCallbacks`.
    case manual
    case disabled
}

/// Static merchant-supplied facts about the app, fixed at build time.
public struct FrakMetadata: Sendable, Hashable {
    public let name: String?
    public let currency: FrakCurrency
    // Nil means "let the backend decide".
    public let lang: FrakLanguage?
    public let logoURL: String?
    public let homepageLink: String?

    public init(
        name: String? = nil,
        currency: FrakCurrency = .eur,
        lang: FrakLanguage? = nil,
        logoURL: String? = nil,
        homepageLink: String? = nil
    ) {
        self.name = name
        self.currency = currency
        self.lang = lang
        self.logoURL = logoURL
        self.homepageLink = homepageLink
    }
}

/// Everything the SDK needs to start, supplied once to `Frak.initialize(_:)`.
public struct FrakConfig: Sendable, Hashable {
    // Optional: when nil, resolved from bundleId instead.
    public let merchantId: String?
    // Nil reads Bundle.main.bundleIdentifier at Frak.initialize(_:).
    public let bundleId: String?
    public let metadata: FrakMetadata
    // Merchants never set this; for Frak's own dev/local builds.
    public let env: FrakEnvironment
    public let deepLink: DeepLinkHandling
    /// Whether tracking may run. `false` means no anonymous id is ever minted and no tracking
    /// request is issued; it is a hard floor that `FrakClient.setTrackingEnabled(_:)` cannot lift
    /// at runtime.
    ///
    /// Not a whole-SDK off switch: merchant config and reward resolution still run, since they
    /// carry no identifier for the user. Sharing does stop, since a share link is the anonymous id.
    ///
    /// Leave it `true` and drive consent through `FrakClient.setTrackingEnabled(_:)` instead
    /// unless you want a build that can never track.
    public let trackingEnabled: Bool
    public let logLevel: FrakLogLevel
    // Nil (default) keeps diagnostics in os.Logger.
    public let logSink: (any FrakLogSink)?

    public init(
        merchantId: String? = nil,
        bundleId: String? = nil,
        metadata: FrakMetadata = FrakMetadata(),
        env: FrakEnvironment = .production,
        deepLink: DeepLinkHandling = .manual,
        trackingEnabled: Bool = true,
        logLevel: FrakLogLevel = .none,
        logSink: (any FrakLogSink)? = nil
    ) {
        self.merchantId = merchantId
        self.bundleId = bundleId
        self.metadata = metadata
        self.env = env
        self.deepLink = deepLink
        self.trackingEnabled = trackingEnabled
        self.logLevel = logLevel
        self.logSink = logSink
    }

    func withBundleId(_ bundleId: String) -> FrakConfig {
        FrakConfig(
            merchantId: merchantId,
            bundleId: bundleId,
            metadata: metadata,
            env: env,
            deepLink: deepLink,
            trackingEnabled: trackingEnabled,
            logLevel: logLevel,
            logSink: logSink
        )
    }

    // logSink can't be Hashable (existential over Sendable-only protocol), so it's
    // excluded from equality/hashing below.
    public static func == (lhs: FrakConfig, rhs: FrakConfig) -> Bool {
        lhs.merchantId == rhs.merchantId && lhs.bundleId == rhs.bundleId && lhs.metadata == rhs.metadata
            && lhs.env == rhs.env && lhs.deepLink == rhs.deepLink && lhs.trackingEnabled == rhs.trackingEnabled
            && lhs.logLevel == rhs.logLevel
    }

    public func hash(into hasher: inout Hasher) {
        hasher.combine(merchantId)
        hasher.combine(bundleId)
        hasher.combine(metadata)
        hasher.combine(env)
        hasher.combine(deepLink)
        hasher.combine(trackingEnabled)
        hasher.combine(logLevel)
    }
}
